build.sbt

```diff
@@ -19,6 +21,12 @@ ThisBuild / tlJdkRelease := Some(8)
 
 ThisBuild / resolvers ++= Resolver.sonatypeOssRepos("snapshots")
 
+lazy val precompile = taskKey[Unit]("runs Calico-specific pre-compile tasks")
+
+precompile := DomDefsGenerator.cachedGenerate()
+
+(Compile / compile) := ((Compile / compile) dependsOn precompile).value
```

Comment:
sbt has a `sourceGenerators` capability for this.

```scala
Compile / sourceGenerators += <task of type Seq[File]>.taskValue
```

Reply:
Laminar also has the generated files checked into source control, but I have no idea why.

Reply:
Laminar has generated sources checked into source control because the fact that they're generated is an implementation detail that the end users don't care about. As a user, I want to see all the code of the library in one place, e.g. I should be able to search the project's github repo to find `idAttr`, `footerTag`, etc. For contributors, the generated sources have a comment on top warning not to edit the files directly, and pointing them to the generator and to SDT. This also essentially makes the generator / SDT an _optional_ dependency, as you can compile Laminar with the generator disabled, if there is ever a need. But mostly I care about UX.

Reply:
> the fact that they're generated is an implementation detail that the end users don't care about

Yeah this is a great perspective and overall I agree. Unfortunately for my users the prospect of checking in generated sources makes me feel queasy 😝

Also I'd argue it's kind of a tooling problem. We upload our source artifacts to maven central, including generated sources. If we can have javadoc.io to browse scaladoc artifacts, then surely we could have a similar service for browsing sources. Also anyone in an IDE would be able to browse the sources too.

Reply:
It helps if you think of the generator as an ~~unpaid intern~~ coding assistant who pumps out a lot of boring repetitive code, as opposed to a macro that is supposed to create opaque code that nobody ever wants to look into :)